Added algorithm for color contrast ratio.

SethFalco opened the pull request:
### **Describe your change:**
This implements the procedure documented on the W3 website to calculate the contrast ratio between two colors.  
https://www.w3.org/TR/WCAG20-TECHS/G17.html#G17-procedure

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

#### References
<!-- Add any reference to previous pull-request or issue -->
N/A

### **Checklist:**
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Java files are placed inside an existing directory.
* [x] All filenames are in all uppercase characters with no spaces or dashes.
* [x] All functions and variable names follow Java naming conventions.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.

SethiPandi commented:
I'm closing this as I've replaced it with https://github.com/TheAlgorithms/Java/pull/2023 which merged to `Development` instead and includes unit tests.

ayaankhan98 commented:
would you please remove `convolutionFFT.java` from this PR, as this file is not related to this PR.

SethFalco commented:
> would you please remove `convolutionFFT.java` from this PR, as this file is not related to this PR.

I could _try_, but that was the Google Java Formatter that did that, not me.
If I undo it, won't it just format it again?

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,96 @@
+import java.awt.Color;
+
+/**
+ * A Java implementation of the offcial W3 documented procedure to calculate contrast ratio between
+ * colors on the web.
+ *
+ * <p>This is used to calculate the readability of a foreground color on top of a background color.
```
```suggestion
 * @brief A Java implementation of the offcial W3 documented procedure to calculate contrast ratio between
 * colors on the web. This is used to calculate the readability of a foreground color on top of a background color.
```

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,96 @@
+ * @since 2020-10-15
+ * @see <a href="https://www.w3.org/TR/WCAG20-TECHS/G17.html#G17-procedure">Color Contrast Ratio
+ *     Procedure</a>
+ * @author Seth Falco (https://elypia.org/)
```
```suggestion
 * @author [Seth Falco](https://github.com/SethFalco)
```

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,96 @@
+ * <p>This is used to calculate the readability of a foreground color on top of a background color.
+ *
+ * @since 2020-10-15
+ * @see <a href="https://www.w3.org/TR/WCAG20-TECHS/G17.html#G17-procedure">Color Contrast Ratio
```
```suggestion
 * @see [Color Contrast Ratio](https://www.w3.org/TR/WCAG20-TECHS/G17.html#G17-procedure)
```

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,96 @@
+ *
+ * @since 2020-10-15
+ * @see <a href="https://www.w3.org/TR/WCAG20-TECHS/G17.html#G17-procedure">Color Contrast Ratio
+ *     Procedure</a>
```
```suggestion
```

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  /**
+   * Calculates the relative luminance of a given color.
```
```suggestion
   * @brief Calculates the relative luminance of a given color.
```

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,96 @@
+  /**
+   * You can check this example against another open-source implementation available on GitHub.
+   *
+   * @see <a
+   *     href="https://contrast-ratio.com/#rgb%28226%2C%20229%2C%20248-on-rgb%2823%2C%20103%2C%20154%29">Online
+   *     Contrast Ratio</a>
```
prefer to use Markdown syntax whereever possible

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,96 @@
+   * @see <a
+   *     href="https://contrast-ratio.com/#rgb%28226%2C%20229%2C%20248-on-rgb%2823%2C%20103%2C%20154%29">Online
+   *     Contrast Ratio</a>
+   * @see <a href="https://github.com/LeaVerou/contrast-ratio">GitHub Repository for Online Contrast
+   *     Ratio</a>
```
prefer to use Markdown syntax whereever possible

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  /**
+   * Calculates the contrast ratio between two given colors.
```
```suggestion
   * @brief Calculates the contrast ratio between two given colors.
```

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,96 @@
+   *
+   * @param color Any color, used to get the red, green, and blue values.
+   * @return The relative luminance of the color.
+   * @see <a href="https://www.w3.org/TR/2008/REC-WCAG20-20081211/#relativeluminancedef">More info
+   *     on relative luminance.</a>
```
prefer to use Markdown syntax whereever possible

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  /**
+   * Calculates the final value for a color to be used in the relative luminance formula as
```
```suggestion
   * @brief Calculates the final value for a color to be used in the relative luminance formula as
```

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  /**
+   * Calculates the Color sRGB value as denoted in step 1 of the procedure document.
```
```suggestion
   * @brief Calculates the Color sRGB value as denoted in step 1 of the procedure document.
```

ayaankhan98 commented:
> > would you please remove `convolutionFFT.java` from this PR, as this file is not related to this PR.
> 
> I could _try_, but that was the Google Java Formatter that did that, not me.
> If I undo it, won't it just format it again?

it is weird, Fromatter will format only those files which are linked to the PR. If it starts formatting files which are not linked in PR then there are so many files in the repository which are not up to standard, it would have formatted all of them.

Well try to remove this from this PR.

SethFalco commented:
I removed both `DIRECTORY.md` and `ConvolutionFFT.java `, however it seems the bot decided to apply the formatting to `ConvolutionFFT.java ` again, and also split up the markdown in the javadocs. :thinking:

ayaankhan98 commented:
> I removed both `DIRECTORY.md` and `ConvolutionFFT.java `, however it seems the bot decided to apply the formatting to `ConvolutionFFT.java ` again, and also split up the markdown in the javadocs. 🤔

How did you removed `ConvolutionFFt.java` from PR?

SethFalco commented:
> How did you removed `ConvolutionFFt.java` from PR?

I just did `git checkout origin/master -- Maths/ConvolutionFFT.java` to revert it back to the version in `master`.

ayaankhan98 commented:
> > How did you removed `ConvolutionFFt.java` from PR?
> 
> I just did `git checkout origin/master -- Maths/ConvolutionFFT.java` to revert it back to the version in `master`.

🤔 try `git restore <path to file>`

SethFalco commented:
> thinking try `git restore <path to file>`

Tried it, however no changes were made, no files were staged, and `git log` doesn't return any new commits. 
It appears to have done nothing.

ayaankhan98 reviewed (commented):
Please add a test function to verify the implementation, rest of the code looks good to me.

SethFalco commented:
> Please add a test function to verify the implementation, rest of the code looks good to me.

Just checking, is there a section for unit tests?
This PR already has a `main` method that should be valid.

```java
public static void main(String args[]) {
  final ColorContrastRatio algorithmImpl = new ColorContrastRatio();

  // Relative Luminance: 0.12215748057375966
  final Color foreground = new Color(23, 103, 154);

  // Relative Luminance: 0.7898468477881603
  final Color background = new Color(226, 229, 248);

  // Contrast Ratio: 4.878363954846178
  final double contrastRatio = algorithmImpl.getContrastRatio(foreground, background);

  System.out.println(contrastRatio);
}
```

Looking at other classes, this appears to be how they're testing it also.

SethFalco commented:
May I also ask, you requested some javadocs have `@brief` prepended to them, but I'm unfamiliar with it, and can't see any other uses of it throughout the repository.

Could you explain why you wanted this or what it does?

ayaankhan98 commented:
> > Please add a test function to verify the implementation, rest of the code looks good to me.
> 
> Just checking, is there a section for unit tests?
> This PR already has a `main` method that should be valid.
> 
> ```java
> public static void main(String args[]) {
>   final ColorContrastRatio algorithmImpl = new ColorContrastRatio();
> 
>   // Relative Luminance: 0.12215748057375966
>   final Color foreground = new Color(23, 103, 154);
> 
>   // Relative Luminance: 0.7898468477881603
>   final Color background = new Color(226, 229, 248);
> 
>   // Contrast Ratio: 4.878363954846178
>   final double contrastRatio = algorithmImpl.getContrastRatio(foreground, background);
> 
>   System.out.println(contrastRatio);
> }
> ```
> 
> Looking at other classes, this appears to be how they're testing it also.

yes, i know it does have a main method. what i want to say is that, create a member function with name `tests`.

```Java
void test() {
      final ColorContrastRatio algorithmImpl = new ColorContrastRatio();
      // test 1
      final Color foreground = new Color(23, 103, 154);
      final Color background = new Color(226, 229, 248);
      final double contrastRatio = algorithmImpl.getContrastRatio(foreground, background);
      assert contrastRation == 4.878363954846178 : "Failed test 1";
      // test 2
      ..............................
      ..............................
      .............................
}
public static void main(String [] args) {
     // call test() here to execute test.
}
```

ayaankhan98 commented:
> May I also ask, you requested some javadocs have `@brief` prepended to them, but I'm unfamiliar with it, and can't see any other uses of it throughout the repository.
> 
> Could you explain why you wanted this or what it does?

Currently this repository is not up to a good standard, I am planing to improve this repository as we did in C++ repository. The reason i am enforcing javadocs style is, with the help of javadocs we can generate documentation for this repository automatically. Other files are not mainted properly, we will be improving them soon.

please have a look here https://thealgorithms.github.io/C-Plus-Plus/  (this is the documentaion of C++ repository algorithms, it is generated automatically using Doxygen)
this is what i am thinking to do with this repo too.

ayaankhan98 commented:
Soon we will be updating the contribution guidelines so that people will follow proper guildelines for each and every PR. Thereby facilitating faster reviews and less burden on Maintainers and Contributors.

SethFalco commented:
> yes, i know it does have a main method. what i want to say is that, create a member function with name `tests`.

Understood, thanks for clarifying, I believe I've done this now.

> please have a look here https://thealgorithms.github.io/C-Plus-Plus/ (this is the documentaion of C++ repository algorithms, it is generated automatically using Doxygen)

Ahh I see, I wasn't familiar with Doxygen before now, I've only ever used the vanilla Javadocs. That would explain the markdown and odd tags I've never seen before.

High hopes this is fine now? Don't be afraid to let me know if there are any other issues.

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,103 @@
+
+    if (aColorLuminance > bColorLuminance)
+      return (aColorLuminance + 0.05) / (bColorLuminance + 0.05);
+    else return (bColorLuminance + 0.05) / (aColorLuminance + 0.05);
```
```suggestion
    return (bColorLuminance + 0.05) / (aColorLuminance + 0.05);
```

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,103 @@
+  /**
+   * @brief Calculates the final value for a color to be used in the relative luminance formula as
+   *     described in step 1.
+   * @param color8Bit The 8-bit repretation of a color component value.
```
```suggestion
   * @param value 8-bit of a representation color component value.
```

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,103 @@
+
+  /**
+   * @brief Calculates the Color sRGB value as denoted in step 1 of the procedure document.
+   * @param color8Bit The 8-bit repretation of a color component value.
```
```suggestion
   * @param color8Bit  8-bit representation of a color component value.
```

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,103 @@
+  /**
+   * @brief Calculates the final value for a color to be used in the relative luminance formula as
+   *     described in step 1.
+   * @param color8Bit The 8-bit repretation of a color component value.
```
```suggestion
   * @param color8Bit 8-bit repretation of a color component value.
```

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,103 @@
+   * @param color8Bit The 8-bit repretation of a color component value.
+   * @return Value for the provided color component to be used in the relative luminance formula.
+   */
+  public double getColor(int value) {
```
```suggestion
  public double getColor(int color8Bit) {
```

ayaankhan98 commented on Misc/ColorContrastRatio.java:
```diff
@@ -0,0 +1,103 @@
+   * @return Value for the provided color component to be used in the relative luminance formula.
+   */
+  public double getColor(int value) {
+    final double sRgb = getColorSRgb(value);
```
```suggestion
    final double sRgb = getColorSRgb(color8Bit);
```

ayaankhan98 reviewed (approved):
LGTM! Thanks for your contribution @SethFalco
